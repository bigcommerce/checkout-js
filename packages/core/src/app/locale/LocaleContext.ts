import { CurrencyService, LanguageService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface LocaleContextType {
    language: LanguageService;
    date?: {
        inputFormat: string;
    };
    currency?: CurrencyService;
}

const LocaleContext = createContext<LocaleContextType | undefined>(undefined);

export default LocaleContext;
