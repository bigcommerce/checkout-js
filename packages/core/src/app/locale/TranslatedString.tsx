import React, { Fragment, FunctionComponent } from 'react';

import withLanguage, { WithLanguageProps } from './withLanguage';

export interface TranslatedStringProps {
    id: string;
    data?: any;
}

const TranslatedString: FunctionComponent<TranslatedStringProps & WithLanguageProps> = ({
    data,
    id,
    language,
}) => (
    <Fragment>
        { language.translate(id, data) }
    </Fragment>
);

export default withLanguage(TranslatedString);
