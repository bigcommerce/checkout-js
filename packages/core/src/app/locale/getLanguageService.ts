import { createLanguageService, LanguageConfig, LanguageService } from '@bigcommerce/checkout-sdk';

import { FALLBACK_LOCALE, FALLBACK_TRANSLATIONS } from './translations';

let languageService: LanguageService | undefined;

export default function getLanguageService(): LanguageService {
    languageService =
        languageService ??
        createLanguageService({
            fallbackLocale: FALLBACK_LOCALE,
            fallbackTranslations: FALLBACK_TRANSLATIONS,
        });

    return languageService;
}

export type InitializeLanguageService = typeof initializeLanguageService;

export function initializeLanguageService(config: LanguageConfig): LanguageService {
    languageService = createLanguageService({
        ...config,
        defaultLocale: config.locale,
        defaultTranslations: config.defaultTranslations,
        fallbackLocale: FALLBACK_LOCALE,
        fallbackTranslations: FALLBACK_TRANSLATIONS,
    });

    return languageService;
}
