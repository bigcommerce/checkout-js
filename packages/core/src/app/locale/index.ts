export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
