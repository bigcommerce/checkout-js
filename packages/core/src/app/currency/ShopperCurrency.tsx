import React, { Fragment, FunctionComponent } from 'react';

import { withCurrency, WithCurrencyProps } from '../locale';

export interface ShopperCurrencyProps {
    amount: number;
}

const ShopperCurrency: FunctionComponent<ShopperCurrencyProps & WithCurrencyProps> = ({
    amount,
    currency,
}) => (
    <Fragment>
        { currency.toCustomerCurrency(amount) }
    </Fragment>
);

export default withCurrency(ShopperCurrency);
