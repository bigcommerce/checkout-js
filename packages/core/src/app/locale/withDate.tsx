import { createInjectHoc, InjectHoc } from '../common/hoc';

import LocaleContext from './LocaleContext';

export interface WithDateProps {
    date: {
        inputFormat: string;
    };
}

const withDate: InjectHoc<WithDateProps> = createInjectHoc(LocaleContext, {
    displayNamePrefix: 'withDate',
    pickProps: (value, key) => key === 'date' && !!value,
});

export default withDate;
