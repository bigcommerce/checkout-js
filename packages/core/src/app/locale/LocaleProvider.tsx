import { CheckoutService, createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState> = {};

    private languageService = getLanguageService();
    private unsubscribe?: () => void;

    private getContextValue = memoizeOne((config?: StoreConfig) => {
        return {
            currency: config ? createCurrencyService(config) : undefined,
            date: config
                ? {
                      inputFormat: config.inputDateFormat,
                  }
                : undefined,
            language: this.languageService,
        };
    });

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig(),
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;

        return (
            <LocaleContext.Provider value={this.getContextValue(config)}>
                {children}
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
