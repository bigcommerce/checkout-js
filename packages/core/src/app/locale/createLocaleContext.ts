import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    const { inputDateFormat } = config;

    return {
        currency: createCurrencyService(config),
        date: {
            inputFormat: inputDateFormat,
        },
        language: getLanguageService(),
    };
}
