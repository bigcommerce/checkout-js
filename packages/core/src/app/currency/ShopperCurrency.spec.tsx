import React from 'react';
import testRenderer from 'react-test-renderer';

import { LocaleContext } from '../locale';
import { getLocaleContext } from '../locale/localeContext.mock';

import ShopperCurrency from './ShopperCurrency';

describe('ShopperCurrency Component', () => {
    const localeContext = getLocaleContext();

    it('renders formatted amount in shopper currency', () => {
        const currency = localeContext.currency;

        jest.spyOn(currency, 'toCustomerCurrency');

        const tree = testRenderer
            .create(
                <LocaleContext.Provider value={localeContext}>
                    <ShopperCurrency amount={10} />
                </LocaleContext.Provider>,
            )
            .toJSON();

        expect(currency.toCustomerCurrency).toHaveBeenCalledWith(10);
        expect(tree).toMatchSnapshot();
    });
});
