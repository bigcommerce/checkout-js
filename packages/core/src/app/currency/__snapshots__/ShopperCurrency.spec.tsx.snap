// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopperCurrency Component renders formatted amount in shopper currency 1`] = `"$11.20"`;
